cairo/ethereum/crypto/alt_bn128.cairo

```diff
@@ -171,6 +171,61 @@ func bnp2_point_at_infinity() -> BNP2 {
     return res;
 }
 
+func bnp2_add{range_check96_ptr: felt*, add_mod_ptr: ModBuiltin*, mul_mod_ptr: ModBuiltin*}(
+    p: BNP2, q: BNP2
+) -> BNP2 {
+    alloc_locals;
+
+    let bnf2_zero = BNF2_ZERO();
+    let x_is_zero = BNF2__eq__(p.value.x, bnf2_zero);
+    let y_is_zero = BNF2__eq__(p.value.y, bnf2_zero);
+    let p_is_infinity = x_is_zero * y_is_zero;
+    if (p_is_infinity != 0) {
+        return q;
+    }
+
+    let x_is_zero_q = BNF2__eq__(q.value.x, bnf2_zero);
+    let y_is_zero_q = BNF2__eq__(q.value.y, bnf2_zero);
+    let q_is_infinity = x_is_zero_q * y_is_zero_q;
+    if (q_is_infinity != 0) {
+        return p;
+    }
+
+    let x_equal = BNF2__eq__(p.value.x, q.value.x);
+    if (x_equal != 0) {
+        let y_equal = BNF2__eq__(p.value.y, q.value.y);
+        if (y_equal != 0) {
+            return bnp2_double(p);
+        }
+        let res = bnp2_point_at_infinity();
+        return res;
+    }
+
+    // Standard case: compute point addition using the formula:
+    // λ = (q.y - p.y) / (q.x - p.x)
+    // x_r = λ^2 - p.x - q.x
+    // y_r = λ(p.x - x_r) - p.y
+
+    // Calculate λ = (q.y - p.y) / (q.x - p.x)
+    let y_diff = bnf2_sub(q.value.y, p.value.y);
+    let x_diff = bnf2_sub(q.value.x, p.value.x);
+    let lambda = bnf2_div(y_diff, x_diff);
+
+    // Calculate x_r = λ^2 - p.x - q.x
+    let lambda_squared = bnf2_mul(lambda, lambda);
+    let x_sum = bnf2_add(p.value.x, q.value.x);
+    let x_r = bnf2_sub(lambda_squared, x_sum);
+
+    // Calculate y_r = λ(p.x - x_r) - p.y
+    let x_diff_r = bnf2_sub(p.value.x, x_r);
+    let lambda_times_x_diff = bnf2_mul(lambda, x_diff_r);
+    let y_r = bnf2_sub(lambda_times_x_diff, p.value.y);
```

Comment:
should this part not be an ec_ops circuit?

Reply:
functions calls are not supported in circuits, here calls to bnf2_sub, div and mul are necessary

Reply:
yes they would need to be inlined